### PR TITLE
next-auth: profile callback on ProviderAuth0Options

### PR DESCRIPTION
**Pull request by cyrus-za**
See https://github.com/nextauthjs/next-auth/pull/1015#issuecomment-761579896

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nextauthjs/next-auth/pull/1015#issuecomment-761579896
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/next-auth/providers.d.ts [hunk 1]
```diff
@@ -181,11 +181,19 @@ interface ProviderGoogleOptions extends GenericObject {
...
+    email: string;
+    picture: string;
+}
+
 interface ProviderAuth0Options extends GenericObject {
     name?: string;
     clientId: string;
     clientSecret: string;
     domain: string;
+    profile?: (profile: Auth0Profile) => GenericObject
```
Comment (lluia) [inline types/next-auth/providers.d.ts:196]: You'll need to add a `;` at the end of this line [to make the CI pass](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/50677/checks?check_run_id=1716617844) 💅🏽

Comment (lluia) [reply types/next-auth/providers.d.ts:196]: Given we're at it, I think we can take advantage and type as well the returned object [from `profile()`](https://github.com/nextauthjs/next-auth/blob/canary/src/providers/auth0.js#L13-L18):
```ts
interface ProfileResult {
  id: string;
  name: string;
  email: string;
  image: string;
}
// ... 
profile?: (profile: Auth0Profile) => ProfileResult
```

**Review by 0ubbe — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution 💚 💯 , two small suggestions 💬

File: types/next-auth/providers.d.ts (hunk 1 above)
Comment (cyrus-za) [reply types/next-auth/providers.d.ts:196]: Thanks for the feedback

ProfileResult can be extended by more than those, hence the GenericObject, but let's do a best of both

```ts
interface ProfileResult {
  id: string;
  name: string;
  email: string;
  image: string;
}
// ... 
profile?: <T extends ProfileResult>(profile: Auth0Profile) => T
```

**Review by 0ubbe — APPROVED** [review]
Thanks 💪🏽

**Comment by lluia** [comment]
The CI is failing because of this [lint rule](https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md). 

Having a second thought, I think your initial suggestion is the best 💯 , to just mark as return result `GenericObject`.
As you mentioned, it's up to the client to decide what [`profile()` returns](https://github.com/nextauthjs/next-auth/blob/canary/src/providers/auth0.js#L13-L18).

**Review by 0ubbe — APPROVED** [review]
Thanks 💯

**Comment by lluia** [comment]
Ready to merge

**Pull request merged**